### PR TITLE
osd: bypass readonly ops when osd full.

### PR DESCRIPTION
**Pull request by majianpeng**
Signed-off-by: Jianpeng Ma <jianpeng.ma@intel.com>

### CODE REVIEW

File: src/osd/ReplicatedPG.cc [hunk 1]
```diff
@@ -1568,20 +1568,22 @@ void ReplicatedPG::do_op(OpRequestRef& op)
     return;
   }
 
+  bool readonly = op->may_read() && !(op->may_write() ||  op->may_cache());
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: This isn't sufficient--they may specify the RWORDERED flag.

**Comment by liewegas** [comment]
Is this fixing an actual bug?

**Comment by liewegas** [comment]
I think Objecter needs to e fixe dtoo:

<pre>
  } else if ((op->target.flags & CEPH_OSD_FLAG_WRITE) &&
	     !(op->target.flags & (CEPH_OSD_FLAG_FULL_TRY |
				   CEPH_OSD_FLAG_FULL_FORCE)) &&
	     (_osdmap_full_flag() ||
	      _osdmap_pool_full(op->target.base_oloc.pool))) {
    ldout(cct, 0) << " FULL, paused modify " << op << " tid "
		  << op->tid << dendl;
    op->target.paused = true;
    _maybe_request_map();
</pre>
should check for RWORDERED.

And on the OSD side, we shoudl discard if RWORDERED *or* it is a write.  I.e., make the conditions match.

**Comment by majianpeng** [comment]
@liewegas . When a ceph cluster full, we don't do any read-object ops. But for filesystem if it full, it can allow read data.
Other question: if osd full, CEPH_OSD_OP_DELETE can't bypass. It need add "forcefull' for rados. But why make the delete w/  forcefull at default?

**Comment by majianpeng** [comment]
@xiaoxichen. From git log, 'rados rm  --force-full' was your patch. Why did not make force-full at default for delete-op? Thanks

**Comment by xiaoxichen** [comment]
@majianpeng ,sorry , could you point me to the code?

**Comment by majianpeng** [comment]
@xiaoxichen. Sorry i saw wrongly.  It wrote by @cxwshawn .

**Comment by athanatos** [comment]
This looks right I think, adding needs-qa.

**Comment by tchaikov** [comment]
tested at http://pulpito.ceph.com/kchai-2016-06-12_09:44:10-rados-wip-kefu-testing---basic-mira/

http://pulpito.ceph.com/kchai-2016-06-12_09:44:10-rados-wip-kefu-testing---basic-mira/254583/ addressed by https://github.com/ceph/ceph-qa-suite/pull/1052

**Pull request merged**